routing/control_tower.go

```diff
@@ -255,6 +261,27 @@ func (p *controlTower) SubscribePayment(paymentHash lntypes.Hash) (
 	return subscriber, nil
 }
 
+// SubscribeAllPayments subscribes to updates for all payments. A first update with
+// the current state of every inflight payment is always sent out immediately.
+func (p *controlTower) SubscribeAllPayments() (*ControlTowerSubscriber, error) {
+	inflightPayments, err := p.db.FetchInFlightPayments()
+	if err != nil {
+		return nil, err
+	}
+
+	subscriber := newControlTowerSubscriber()
+	for _, payment := range inflightPayments {
+		// Always write current payment state to the channel.
+		subscriber.queue.ChanIn() <- payment
+	}
+
+	p.subscribersMtx.Lock()
```

Comment:
You may want to think about race conditions here. If a new payment becomes pending right at this point, it will probably be missed on the stream?

Reply:
Yeah, I thought about race conditions here. It's basically a tradeoff between three things:
1) Ignore inflight payments that started before the subscription
2) Do include the inflight payments, but maybe miss an update in between around the start of the subscription (the way it's coded now)
3) Do include the inflight payments and all updates in between, but potentially out of order at the start of the subscription and with potential duplicates. (if we'd add the subscription to the list before fetching inflight payments)

I preferred number 2 myself, that's why it's coded like this.

Reply:
I think that 3 is the ideal option, closest to what the user would expect. Maybe it is worth thinking about this a bit more, especially because there are no missed updates for the single payment subscription (facilitated by `paymentsMtx`).

Maybe a similar solution with a lock is possible for `SubscribeAllPayments`? The updates go into a memory queue, so it won't block on rpc clients not being responsive.

Reply:
I've made an attempt to fix that. Judging from the comments in the code, the payments mutex is there to to do the exact same thing I'm trying to do. But since this change cannot lock on a specific payment hash, we'd need some other locking mechanism.

_If_ the payments mutex is there to avoid payment updates from happening _while_ we're setting up a new subscription, we can achieve a pretty efficient mechanism with a reader-writer lock, as I did in my latest commit. 

A problem could be that the underlying code to update the payments in the database is actually relying on the mutex on the specific payment hash, so no 2 database updates can actually happen at the same time. That's a side-effect of the existing mutex, but something could rely on it. 

In this new scheme, no database updates can happen _while_ we're making a new subscription, but during normal flow, database updates could occur in parallel.

What do you think?

Another option would be 2 mutexes. One for the payment hash (to avoid concurrent database updates) and a reader-writer mutex while taking the subscription (to avoid updates while making the subscription). But it could be hard to follow what's going on as a reader of the code then.

Reply:
I am not sure if it is safe. With only reader locks, I think the order in which db updates are applied isn't guaranteed to be reflected in the subscriber events for a single payment. Maybe if every payment is only updated from a single thread, it is not a problem.

An additional mutex that just prevents events from being missed may be the safer option.

Nice analysis btw.

Reply:
Phew, lots of locks. I went through the code, but have to say that it isn't easy to see how all these combinations of locks, rlocks and unlocks work together and that it is 100% correct. Part of that is pre-existing, it wasn't simple to start with.

Not sure what to do about it though. It may be good to take another look at what exactly the critical path is that we need to optimize for. The expensive operations that prevent us from just wrapping everything in a single lock that kills all concurrency. And from there, see if there are other options.

Reply:
Yeah it makes it very difficult to read indeed.
It could be better to 
- drop the additional locking
- Add the subscriber before fetching the inflight payments

That way the user might get some duplicate events and potentially out of order events at the beginning of the stream, but it saves a lot of complexity in the code here.

Another option is I can merge that subscribersMtx and newSubscriptionMtx maybe. I'll have a look at the exact tradeoffs there.

Reply:
Yes, maybe it is not worth the extra complexity. This might be a good time for another reviewer to take a look at it. It seems the options are quite well laid out now.